prow/cmd/build/controller.go

```diff
@@ -212,6 +212,14 @@ func newController(opts controllerOptions) (*controller, error) {
 	return c, nil
 }
 
+// clusterToCtx converts the prow job's cluster to a cluster context
+func clusterToCtx(cluster string) string {
+	if cluster == kube.DefaultClusterAlias {
+		return kube.InClusterContext
```

Comment:
Jobs do not always run in cluster by default (they do not for prow.k8s.io)

Reply:
Alright, but I think the issue is when no `DefaultClusterAlias` is explicitly defined in the build configs because of this https://github.com/kubernetes/test-infra/blob/126cb79be5652f502c71a99e4fbe221570b7085b/prow/kube/config.go#L131

In this case both configs point to the same cluster, and because an even handler is added for each config:  https://github.com/kubernetes/test-infra/blob/126cb79be5652f502c71a99e4fbe221570b7085b/prow/cmd/build/controller.go#L196

They will end up deleting each other build resources during the reconciliation.

Reply:
Yeah I agree there is a problem, but I am not convinced this solution fully solves the problem.

How will this behave when the default context is not the local cluster?

Reply:
Agree, It's breaking the other scenario. What about if we add the cluster alias in the config map only when it is a different cluster. In the controller, we can add it to the config only when is defined in the map https://github.com/kubernetes/test-infra/blob/c829a465be69a5cee97ec4183e002645275f0a70/prow/cmd/build/main.go#L164. and do the mapping to in cluster context only when is not present in the map.

Reply:
I reworked the fix as explained above. I am not sure if this is going to break something else.

Reply:
This still seems to assume that the default cluster alias and in cluster alias are the same? These can be different things.

AKA the "" alias refers to the same cluster where prow lives. The "default" alias refers to the cluster to schedule resources into when jobs to not set one explicitly.

By default, the default and in-cluster aliases are the same, but they are not always (prow.k8s.io runs prow binaries in cluster X and jobs in cluster Y by default).

Reply:
This makes the assumption that as long as the `default` alias is not explicitly defined in the configuration, then it is the same like the `in-cluster` context and skips adding it to the configs map.

In the same time when `default` alias is not present in the configs map, it won't run the reconciliation for both because they are referring to the same cluster, and they will end up deleting each other builds. 

Let's see some examples: 
`prow.k8s.io` when `default` alias is configured to point to cluster Y:
default => cluster Y
in-cluster => cluster X
both contexts are in the build configs
reconciliation runs for both contexts
prow jobs created with `spec.Cluster: default` are mapped to `default` context which point to cluster Y

default installation when no `default` alias is configured:
in-cluster => cluster X
default => cluster X
only the in-cluster context is in the build configs
reconciliation runs only for in-cluster context 
prow jobs created with `spec.Cluster: default` are mapped to `in-cluster` context instead

Reply:
> the assumption that as long as the default alias is not explicitly defined in the configuration

I do not think this is a valid assumption. The alias "default" always exists. Other components depend on its existence. It may or may not refer to the in cluster config.

The last time we tried something like this we wound up with a bunch of problems, including unsafely scheduling jobs into the wrong cluster, this caused a bunch of pain and rollbacks.

I will be a lot more comfortable when a) you're only changing code in the build controller (not the shared kube package) and b) you're mapping "" to "default" rather than the other way around.

Reply:
Got it. I will map the in-cluster to default alias only in the build controller. This also fix the issue.

Reply:
Now the in-cluster context is mapped to default alias.